### ISSUE TITLE
SG_NEXT-RECORD empty box not cleaned up after merge

### ISSUE THREAD

**Issue by rotor-head** [berteh/ScribusGenerator#127]
I modified a template file and updated from the prior syntax to the new syntax (along with a fresh download of the script.)

Running
```
 ./ScribusGeneratorCLI.py /Users/me/Documents/placemat.sla -c "/Users/me/Documents/Chapter Members.csv" -n placemat-merged -o /Users/me/Documents/ -m -s
```
There are 27 records in the CSV.
The rows were duplicated using the 'Multiple Duplicate' function in Scribus.
The %SG_Next-Record% is appended to each row.
The final row has the %SG_Next-Record% removed.

What I'm expecting is a single page with all the data from each row of the CSV in the form of a list.

What is actually occurring is each record is written out 26 times on each page with the %SG_Next-Record% at the end of the row and a page for each record created (26 pages.)

The was working with the prior %VAR_Next-Record%, so I must be missing something. Syntax perhaps? I'm a bit perplexed!

Running Scribus 1.5.3

**Comment by rotor-head** [comment]
I found a build of 1.5.4 available on SourceForge and tested.

It works *almost* as expected. It's only missing the first record in the set. The CSV has a header row. Not quite sure how it's skipping the 1st line of data and would appreciate a note on fixing that issue but it's a relief everything else is functioning as designed.

Short answer to the bigger problem, pull the build 1.5.4 presuming that 2.8 of the Generator won't be backward compatible or perhaps, use the prior version of SG with 1.5.3.

**Comment by berteh** [comment]
Looks to me like the ''./ScribusGeneratorCLI.py'' from your command line
has not been updated. Would you have an 'old' version in the Path? or
linked to?

I'll check it out anyway.
Have a nice day.

**Comment by rotor-head** [comment]
I've not added it to the shell path. Having to go the hard way and cd into the directory in order to run the scripts; even went so far as to invoke the help screen to double check the syntax. That's what lead me to search for 1.5.4. Perhaps that will save you some time!

**Comment by berteh** [comment]
As for the first record missing it's due to the implementation of next
record: as soon as SG detects the token it loads the new values from CSV,
instead of finishing the replacing with current ones.

As I found no better way I recommend you put the %SG_Next-Record% token at
the beginning of the second line (and the next ones as well), rather than
at the end of the line/frame/block.

The internal order of elements changes with groups, layers and elements
depth as well, so you may need to play a bit with these in case of a very
complex template.

Any idea how to make it more elegant and user friendly is welcome.

**Comment by berteh** [comment]
Maybe... but that is unlikely the issue, since the CLI version of SG is
completely independent from Scribus... and directly works on the XML.

**Comment by rotor-head** [comment]
Guessing it was all the mucking around with the 1.5.3 macOS build I did a few months ago attempting to get the GUI to work. It must have caused an issue with the xml output.

Placing the ```%SG_NEXT-RECORD%``` at the beginning of the second row is the correct way.

I suspect each use case is going to have special variants. I'm happy to write up my particular use case for you to consider including in the 'examples' directory.

**Comment by InnocentLunatic** [comment]
I'm also working in Scribus 1.5.4 with a freshly downloaded script and New Record behaves bizarrely at best.  I have 4 groups in my 2 page document, each one meant to reference a line in my excel spreadsheet.  I've spent a few hours trying to figure out how it determines which elements to fill first, and I have only gotten more baffled as I have now discovered that it will sometimes even fill a later page before finishing the first one.  I've made some efforts to get it to behave correctly, such as making groups, making sure the first groups have lower numbers, even renaming the text boxes with the next record key to be AText# or AAText#. Also, it will regularly put things from different lines in my excel document within the same group, ex 3-4 items will have the name changed correctly, but will all have the same data for another variable rather than reading the spreadsheet properly.  Sometimes it skips lines altogether, and sometimes it will somehow only copy part of the data cell (I had one name that was 2 words long and for whatever reason it only brought the second word and not the first)

Could you maybe give some more guidance on how it decides which elements to populate first to take the guesswork out of it? I.E Is group most important, is alphabetical order of the name most important, does it prioritize image frames over text boxes or vice versa?  I've uploaded my files in case you want to look at them.
[arbres.csv.txt](https://github.com/berteh/ScribusGenerator/files/2949894/arbres.csv.txt)
[generator experiment.sla.txt](https://github.com/berteh/ScribusGenerator/files/2949895/generator.experiment.sla.txt)

**Comment by rotor-head** [comment]
I've had to take my Excel generate .csv files and change the formatting from Microsoft's format to utf-8 in order for clean imports. In my case, I have records order by lowest to highest and these must be sequential. The merges were always failing when attempting to use a straight Excel .csv. I use a Mac specific app to modify the format that I use for merging.

Since you appear to be on a Windows machine, you can install Libre Office and by-pass the pain introduced by Excel. Of course, your milage may vary.

**Comment by InnocentLunatic** [comment]
I use Linux and my .csv's were made in Libre Office... However, after playing around I am very very pleased to say I found a solution that worked for me, and will hopefully work for you as well if you didn't already solve your problem.

In the save as csv settings, telling Libreoffice to put all fields in quotes seems to have fixed some clipped data issues.

In Scribus, the Scribus generator script prioritizes items based on their Level in Scribus.  This means that your first group should be on the first level.  It also means that the New Record tag also needs to in the first level of the items you're grouping before you group them for it to behave properly (assuming you call new record at the beginning of your 2nd, etc items).  Only bizarre behavior left now is that if you put %SG_New-Record% in it's own text box, the script will not delete the text box like it does with blank variables, but it will at least leave the text box blank.

**Comment by rotor-head** [comment]
Just curious @berteh - is the syntax (specifically case) supposed to matter here? 

My first template I used ``%SG_Next-Record%`` and it created a single page output for each record (which wasn't what I expected.)

The second template I used all caps ``%SG_NEXT-RECORD%`` and it created a single page with all records nicely in a row. This case is noted in the README.md.

It could be that I just didn't read the fine print and it's documented. But just in case it's missing, I  thought I'd ask considering some of the issues. Perhaps it's assumed knowledge for those familiar with Python scripting.

**Comment by berteh** [comment]
indeed @rotor-head all tokens (%VAR... and %SG...) are case sensitive. And you're right to mention it may no be mentionned enough in the documentation, I just reworded it a bit in the README, hoping to makes it clear.


@InnocentLunatic thanks for the detailed feedback, I added your explanations on priority and groups in the README to (hopefully) help others. I'll check out the empty box not getting cleaned bug.

Examples are always really welcome, so if you @InnocentLunatic want to "anonymize" your files by replacing personnal data by random text and put a link to them on the "gallery and examples" wiki that'd be great: https://github.com/berteh/ScribusGenerator/wiki
